flink-runtime/src/main/java/org/apache/flink/runtime/operators/chaining/ChainedReduceCombineDriver.java

```diff
@@ -0,0 +1,305 @@
+import org.apache.flink.runtime.operators.DriverStrategy;
+import org.apache.flink.runtime.operators.hash.ReduceHashTable;
+import org.apache.flink.runtime.operators.sort.FixedLengthRecordSorter;
+import org.apache.flink.runtime.operators.sort.InMemorySorter;
+import org.apache.flink.runtime.operators.sort.NormalizedKeySorter;
+import org.apache.flink.runtime.operators.sort.QuickSort;
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.MutableObjectIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Chained version of ReduceCombineDriver.
+ */
+public class ChainedReduceCombineDriver<T> extends ChainedDriver<T, T> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(ChainedReduceCombineDriver.class);
+
+	/** Fix length records with a length below this threshold will be in-place sorted, if possible. */
+	private static final int THRESHOLD_FOR_IN_PLACE_SORTING = 32;
+
+
+	private AbstractInvokable parent;
+
+	private TypeSerializer<T> serializer;
+
+	private TypeComparator<T> comparator;
+
+	private ReduceFunction<T> reducer;
+
+	private DriverStrategy strategy;
+
+	private InMemorySorter<T> sorter;
+
+	private QuickSort sortAlgo = new QuickSort();
+
+	private ReduceHashTable<T> table;
+
+	private List<MemorySegment> memory;
+
+	private volatile boolean canceled;
+
+	// ------------------------------------------------------------------------
+
+	@Override
+	public Function getStub() {
+		return reducer;
+	}
+
+	@Override
+	public String getTaskName() {
+		return taskName;
+	}
+
+	@Override
+	public void setup(AbstractInvokable parent) {
+		this.parent = parent;
+		canceled = false;
+
+		strategy = config.getDriverStrategy();
+
+		reducer = BatchTask.instantiateUserCode(config, userCodeClassLoader, ReduceFunction.class);
+		FunctionUtils.setFunctionRuntimeContext(reducer, getUdfRuntimeContext());
+	}
+
+	@Override
+	public void openTask() throws Exception {
+		// open the stub first
+		final Configuration stubConfig = config.getStubParameters();
+		BatchTask.openUserCode(reducer, stubConfig);
+
+		// instantiate the serializer / comparator
+		serializer = config.<T>getInputSerializer(0, userCodeClassLoader).getSerializer();
+		comparator = config.<T>getDriverComparator(0, userCodeClassLoader).createComparator();
+
+		MemoryManager memManager = parent.getEnvironment().getMemoryManager();
+		final int numMemoryPages = memManager.computeNumberOfPages(config.getRelativeMemoryDriver());
+		memory = memManager.allocatePages(parent, numMemoryPages);
+
+		LOG.debug("ChainedReduceCombineDriver object reuse: " + (objectReuseEnabled ? "ENABLED" : "DISABLED") + ".");
+
+		switch (strategy) {
+			case SORTED_PARTIAL_REDUCE:
+				// instantiate a fix-length in-place sorter, if possible, otherwise the out-of-place sorter
+				if (this.comparator.supportsSerializationWithKeyNormalization() &&
```

Comment:
You can remove `this` from the copied code. Will be more consistent within this class.